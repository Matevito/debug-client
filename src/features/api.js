const axios = require('axios');

const api = axios.create({
    baseURL: "https://pure-falls-26749.herokuapp.com/apiv1/",
    timeout: 3000
});

export default api;