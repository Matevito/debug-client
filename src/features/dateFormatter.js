import { DateTime } from "luxon"

const simpleDate = (string) => {
    return DateTime.fromISO(string).toLocaleString(DateTime.DATE_MED)
};

const detailedDate = (string) => {
    return DateTime.fromISO(string).toLocaleString(DateTime.DATETIME_FULL_WITH_SECONDS)
}

export {
    simpleDate,
    detailedDate
}