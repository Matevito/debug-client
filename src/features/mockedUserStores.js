import configureStore from "redux-mock-store";

// todo: set project an issues values
const usersStoreValues = [
    {
        user: {} 
    },
    {
        user: {
            loggedIn: true,
            token : "a token?",
            user: {
                username: "devUser",
                email:"dev@email.com",
                role: "Developer",
                id: "1"
            },
            projects: {
                number: 0,
                list: []
            },
            issues: {
                number: 0,
                list: []
            }
        }
    },
    {

        user: {
            loggedIn: true,
            token : "a token?",
            user: {
                username: "TeamLeaderUser",
                email:"team@email.com",
                role: "Team leader",
                id: "2"
            },
            projects: {
                number: 0,
                list: []
            },
            issues: {
                number: 0,
                list: []
            }
        }
    },
    {

        user: {
            loggedIn: true,
            token : "admintoken?",
            user: {
                username: "adminUser",
                email:"admin@email.com",
                role: "Admin",
                id: "3"
            },
            projects: {
                number: 0,
                list: []
            },
            issues: {
                number: 0,
                list: []
            }
        }
    }
];

exports.getData = () => {
    const mockedUsers = usersStoreValues.map((user, i) => {
        const mockStore = configureStore([]);
        if (i === 0) {
            return mockStore(user)
        }
        return mockStore({user: {
            user: {
                user
            }
        }})
    });

    return mockedUsers;
}