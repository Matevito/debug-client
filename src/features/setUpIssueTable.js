const setUpIssueTable = (issuesList) => {
    let data = [
        {label: "open", x: 1,  y: 0},
        {label: "aditional info needed", x: 2, y: 0},
        {label: "in progress", x: 3, y: 0},
        {label: "under review", x: 4, y: 0},
        {label: "solved",x: 5, y: 0}
    ];

    issuesList.forEach((issue) => {
        data.forEach((obj, index) => {
            if (obj.label === issue.status){
                data[index].y += 1
            }
        })
    });
    
    return data.filter(obj => obj.y > 0);
};


export default setUpIssueTable;