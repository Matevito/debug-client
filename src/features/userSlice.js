import  { createSlice } from "@reduxjs/toolkit";

/* User object structure
user:{
    loggedIn: 'boolean',
    user: {
        username,
        email,
        role,
        id
    },
    projects: [{
        id,
        title
    }],
    issues: [{
        id,
        title
    }],
    todo on it's own slice:?
    notifications: [array of id's] // under development*
}
*/
export const userSlice = createSlice({
    name: 'user',
    initialState: {
        user: null
    },
    reducers: {
        login: (state, action) => {
            state.user = action.payload
        },
        logout: (state) => {
            state.user = null
        }
    }
});

export const { login, logout } = userSlice.actions;

export const selectUser = (state) => state.user.user;

export default userSlice.reducer