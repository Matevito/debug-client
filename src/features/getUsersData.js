import api from "./api";

const getUserData = async (token) => {
    const config = {
        headers: { "auth-token" : token}
    }; 
    try {
        const apiRes = await api.get("/user/list", config);
        return apiRes.data.data
    } catch(err) {
        return null
    }
}

export default getUserData