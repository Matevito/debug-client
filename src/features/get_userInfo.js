import api from "./api";

const get_userInfo = async (token) => {
    // setting up headers for calls
    const config = {
        headers: {
            "auth-token": token
        }
    };
    // fetch api data
    const userData = await api.get("/whoami", config);
    if (!userData) {
        return false;
    }
    const userId = userData.data.data.id;
    const userRelatedData = await api.get(`/user/${userId}`, config);
    if (!userRelatedData) {
        return false;
    };

    const userNotifications = [];

    // structure res and send it
    const dataRes = {
        loggedIn: true,
        token: token,
        user: {
            username: userData.data.data.username,
            email: userData.data.data.email,
            role: userData.data.data.role,
            id: userData.data.data.id
        },
        projects: userRelatedData.data.data.projects,
        issues: userRelatedData.data.data.issues,
        notifications: {
            list: userNotifications,
            number: userNotifications.length
        }
    };
    return dataRes;
}

export default get_userInfo;