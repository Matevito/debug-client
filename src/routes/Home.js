import React, { useEffect, useState } from 'react'
import { useNavigate } from "react-router-dom";
import { VictoryPie } from 'victory';

// redux state management
import { useSelector } from "react-redux";
import { selectUser } from "../features/userSlice";

// mui comp
import {
    Box,
    Card, 
    CardContent,
    Typography,
    Grid,
} from "@mui/material";

// Home components
import { Welcome } from "../components/Welcome";
import { LoadingPage } from "../components/LoadingPage";
import { ProjectsTable } from "../components/ProjectsTable";
import { UserTicketsTable } from '../components/UserTicketsTable';

import setUpIssueTable from "../features/setUpIssueTable";

// api comp
import api from "../features/api";

export const Home = () => {
    const user = useSelector(selectUser);
    let navigate = useNavigate();

    const [authorized, setAuthorized] = useState(null);
    const [projectList, setProjectList] = useState([]);

    useEffect(() => {
        async function getProjectList() {
            const config = {
                headers: { "auth-token": user.token }
            };
            const url = "/project/list";

            try {
                const listRes = await api.get(url, config);
                setProjectList(listRes.data.data);
                setAuthorized(true);
            } catch(err) {
                setAuthorized(false);
                navigate("/protected-route");
            }
        };

        if (user){
            // set-up data;
            if (user.user.role !== "Admin"){
                setAuthorized(true)
            } else {
                // fetch projectList info for admin view!
                getProjectList()
            }
        }
    }, [user, navigate])

    if (!user) {
        return(
            <Welcome />
        )
    } else if (authorized === false){
        return (
            <LoadingPage />
        )
    } else if (authorized === true) {
        return (
            <>
            <Card sx={{ width: "100%"}}>
                    <CardContent sx={{ backgroundColor: "#A9A9A9"}}>
                        <Typography variant="h6">Logged as {user.user.role}</Typography>
                    </CardContent>
            </Card>
            <Box
                fullWidth
                sx={{
                    display: "flex",
                    flexDirection: "column",
                    alignItems: "center",
                    hieght: "100vh",
                    m: 2
                }}
            >
                {user.user.role === "Admin" ? <>only admin</> : <></>}
                {user.user.role !== "Developer" ? <>team leader and admin</> : <></>}
                <Card raised={true} sx={{width: "1200px", marginBottom: "20px"}}>
                    <CardContent
                        sx={{
                            backgroundColor:"#C0C0C0"
                        }}
                    >
                        <Grid container spacing={2} sx={{m:2}}>
                            <Grid item xs={6} 
                                sx={{
                                    display: "flex",
                                    flexDirection: "column",
                                    alignItems: "center",
                                    justifyContent: "center"
                                }}
                            >
                                {user.projects.number > 0 ? 
                                    <ProjectsTable  projects={user.projects.list} title={"USER PROJECTS"}/>
                                    : 
                                    <Typography variant="h5">You are not currently assigned to a project. Comunicate with an admin to be assigned to a project!</Typography>
                                }
                                
                            </Grid>
                            <Grid item xs={6} 
                                sx={{
                                    display: "flex",
                                    flexDirection: "column",
                                    alignItems: "center",
                                    justifyContent: "center"
                                }}
                            >
                                <Typography variant="h4"> User projects: {user.projects.number}</Typography>
                            </Grid>
                        </Grid>
                    </CardContent>
                </Card>

                <Card raised={true} sx={{width: "1200px", marginBottom: "20px"}}>
                    <CardContent
                        sx={{
                            backgroundColor:"#C0C0C0"
                        }}
                    >
                        <Grid container spacing={2} sx={{m:2}}>
                            <Grid item xs={5}>
                                <Typography variant="h4">Assigned tickets: {user.issues.list.filter(i => i.status !== "solved").length}</Typography>
                                <Typography variant="h4">Solved tickets: {user.issues.list.filter(i => i.status === "solved").length}</Typography>
                                <Box sx={{
                                    display: "flex",
                                    flexDirection: "column",
                                    m: 2
                                }}>
                                    {user.issues.number > 0 ? 
                                        <VictoryPie
                                            innerRadius={50}
                                            data={setUpIssueTable(user.issues.list)}
                                        />
                                    : <></>
                                    }
                                </Box>
                            </Grid>
                            <Grid item xs={6} sx={{
                                display: "flex",
                                flexDirection: "column",
                                alignItems: "center",
                                justifyContent: "center"
                            }}>
                                {user.issues.list.filter(i => i.status !== "solved").length > 0 ?
                                    <UserTicketsTable issues={user.issues.list} projects={user.projects.list}/>
                                    :
                                    <Typography variant="h5">You don't have tickets assigned. Take one from one of the projects you're into.</Typography>
                                }
                            </Grid>
                        </Grid>
                    </CardContent>
                </Card>
            </Box>
            </>
        )
    }
}
