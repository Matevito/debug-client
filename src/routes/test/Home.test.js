import { Provider } from "react-redux";
import { MemoryRouter, Route, Routes } from "react-router-dom";

// testing comps
import { render, screen, waitFor } from "@testing-library/react";
import '@testing-library/jest-dom'

// test component
import { Home } from "../Home";

// mock functions
import { getData } from "../../features/mockedUserStores";
import { rest } from "msw";
import { setupServer } from "msw/node";

const rootAPI = "https://pure-falls-26749.herokuapp.com/apiv1";

const server = setupServer(
    rest.get("*", (req, res, ctx) => {
        return res(ctx.status(400), ctx.json({ error: "error fetching data"}))
    })
);
const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUsedNavigate,
}));

const renderComponent = (store) => {
    return render(
        <Provider store={store}>
            <MemoryRouter initialEntries={["/"]}>
                <Routes>
                    <Route 
                        path="/"
                        element={<Home />}
                    />
                </Routes>
            </MemoryRouter>
        </Provider>
    )
}
describe("Home component", () => {
    let noLoggedUser;
    let devStore;
    let teamLeaderStore;
    let adminStore;

    beforeEach(() => {
        const usersData = getData();
        noLoggedUser = usersData[0];
        devStore = usersData[1];
        adminStore = usersData[3];
    })
    // required to mock rest-api server
    beforeAll(() => server.listen())
    afterEach(() => server.resetHandlers());
    afterAll(() => {
        server.close()
        jest.clearAllMocks();
    })

    // auth tests
    test.todo("handles user not logged-in");
    test("handles error fetching admin user", async() => {
        renderComponent(adminStore);
        await waitFor(() => mockedUsedNavigate.mock.lastCall[0]);
        expect(mockedUsedNavigate).toHaveBeenCalledWith("/protected-route")
    });

    // functional tests
    test("handles developer", async() => {
        const { container } = renderComponent(devStore);
        await screen.findByText("Logged as Developer");
        expect(container).toMatchSnapshot()
    });
    test.todo("render team leader table");
    test.todo("renders admin user tables");
})