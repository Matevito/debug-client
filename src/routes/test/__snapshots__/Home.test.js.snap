// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component handles developer 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-tjaq38-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-ny9utc-MuiCardContent-root"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
      >
        Logged as 
        Developer
      </h6>
    </div>
  </div>
  <div
    class="MuiBox-root css-1u3twe0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiCard-root css-k363oc-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardContent-root css-94kbed-MuiCardContent-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-3unm91-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1k2cwzv-MuiGrid-root"
          >
            <div
              class="MuiTableContainer-root css-1ofbc68-MuiTableContainer-root"
            >
              <table
                class="MuiTable-root css-1w3qztj-MuiTable-root"
              >
                <thead
                  class="MuiTableHead-root css-1lhsthh-MuiTableHead-root"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                      colspan="6"
                      scope="col"
                    >
                      USER PROJECTS
                    </th>
                  </tr>
                  <tr
                    class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                      colspan="2"
                      scope="col"
                    >
                      PROJECT
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                      scope="col"
                    >
                      TEAM LEADER
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                      scope="col"
                    >
                      ISSUES
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                      scope="col"
                    >
                      ISSUES SOLVED
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                      scope="col"
                    >
                      LINK
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="MuiTableBody-root css-n1y5x-MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                      colspan="2"
                    />
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
                    >
                      <a
                        href="/user/undefined"
                        style="text-decoration: none;"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="AccountCircleIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
                          />
                        </svg>
                      </a>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
                    />
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
                    />
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
                    >
                      <a
                        href="/project/testProject1"
                        underline="hover"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="BuildIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="m22.7 19-9.1-9.1c.9-2.3.4-5-1.5-6.9-2-2-5-2.4-7.4-1.3L9 6 6 9 1.6 4.7C.4 7.1.9 10.1 2.9 12.1c1.9 1.9 4.6 2.4 6.9 1.5l9.1 9.1c.4.4 1 .4 1.4 0l2.3-2.3c.5-.4.5-1.1.1-1.4z"
                          />
                        </svg>
                      </a>
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                      colspan="2"
                    />
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
                    >
                      <a
                        href="/user/undefined"
                        style="text-decoration: none;"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="AccountCircleIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
                          />
                        </svg>
                      </a>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
                    />
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
                    />
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
                    >
                      <a
                        href="/project/testProject2"
                        underline="hover"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="BuildIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="m22.7 19-9.1-9.1c.9-2.3.4-5-1.5-6.9-2-2-5-2.4-7.4-1.3L9 6 6 9 1.6 4.7C.4 7.1.9 10.1 2.9 12.1c1.9 1.9 4.6 2.4 6.9 1.5l9.1 9.1c.4.4 1 .4 1.4 0l2.3-2.3c.5-.4.5-1.1.1-1.4z"
                          />
                        </svg>
                      </a>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1k2cwzv-MuiGrid-root"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
            >
               User projects: 
              2
            </h4>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiCard-root css-k363oc-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardContent-root css-94kbed-MuiCardContent-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-3unm91-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
            >
              Assigned tickets: 
              0
            </h4>
            <h4
              class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
            >
              Solved tickets: 
              0
            </h4>
            <div
              class="MuiBox-root css-1muirhe"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1k2cwzv-MuiGrid-root"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              You don't have tickets assigned. Take one from one of the projects you're into.
            </h5>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
